jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/TransitNodeRoutingShortestPath.java

```diff
@@ -0,0 +1,264 @@
+ * is a shortest path between the vertices which contains a transit vertex. Therefore the
+ * forward access vertices of $u$ and backward access vertices of $v$ are inspected to find
+ * a pair of such access vertices $(a_u, a_v)$ so that the value of $d(u,a_u) + d(a_u, a_v)
+ * + d(a_u, v)$ is minimum over all such pairs. Here $d(s,t)$ is the distance from vertex
+ * $s$ to vertex $t$.
+ *
+ * <p>
+ * The algorithm is designed to operate on sparse graphs with low average outdegree. Comparing
+ * to {@link ContractionHierarchyBidirectionalDijkstra} it uses significantly more time on the
+ * precomputation stage. Because of that it makes sense to use this algorithm on large instances
+ * (i.e. with more than 10.000 vertices), where it shows substantially better performance results
+ * than {@link ContractionHierarchyBidirectionalDijkstra}. Typically this algorithm is used as the
+ * backend for large scale shortest path search engines, e.g.
+ * <a href="https://www.openstreetmap.org">OpenStreetMap</a>.
+ *
+ * @param <V> graph vertex type
+ * @param <E> graph edge type
+ * @author Semen Chudakov
+ * @see TransitNodeRoutingPrecomputation
+ * @see BidirectionalDijkstraShortestPath
+ */
+public class TransitNodeRoutingShortestPath<V, E> extends BaseShortestPathAlgorithm<V, E> {
+
+    /**
+     * Contraction hierarchy which is used to compute shortest paths.
+     */
+    private ContractionHierarchy<V, E> contractionHierarchy;
+
+    /**
+     * Fallback shortest path algorithm for local queries.
+     */
+    private ShortestPathAlgorithm<V, E> localQueriesAlgorithm;
+
+    /**
+     * Many-to-many shortest paths between transit vertices.
+     */
+    private ManyToManyShortestPaths<V, E> manyToManyShortestPaths;
+    /**
+     * Stores access vertices for each vertex in the {@code graph}.
+     */
+    private AccessVertices<V, E> accessVertices;
+    /**
+     * Locality filter which is used to determine if two vertices in the graph
+     * are local to each other or not.
+     */
+    private LocalityFilter<V> localityFilter;
+
+    /**
+     * Constructs a new instance for the given {@code graph}.
+     *
+     * @param graph graph
+     */
+    public TransitNodeRoutingShortestPath(Graph<V, E> graph) {
+        super(graph);
+    }
+
+    /**
+     * Constructs a new instance of the algorithm for a given {@code transitNodeRouting}.
+     *
+     * @param transitNodeRouting transit node routing for {@code graph}
+     */
+    TransitNodeRoutingShortestPath(TransitNodeRouting<V, E> transitNodeRouting) {
+        super(transitNodeRouting.getContractionHierarchy().getGraph());
+        initialize(transitNodeRouting);
+    }
+
+    /**
+     * Lazy computes {@code TransitNodeRouting} for this algorithm. If not called directly
+     * this method will be invoked in ether if {@code getPath()} or {@code getPathWeight()} methods.
+     */
+    public void initializeAlgorithm() {
```

Comment:
I would call this method `performPrecomputation()` rather than `initializeAlgorithm()` because initialization is performed in the constructor. Also, please state that `TransitNodeRouting` is the result of the precomputation stage, cause otherwise, it's not clear. `TransitNodeRouting` is a package-private static class in another package-private class.